src/uct/tcp/tcp_ep.c

```diff
@@ -460,9 +461,14 @@ void uct_tcp_ep_destroy(uct_ep_h tl_ep)
         /* purge all outstanding operations (GET/PUT Zcopy, flush operations) */
         uct_tcp_ep_purge(ep, UCS_ERR_CANCELED);
         uct_tcp_cm_insert_ep(iface, ep);
-    } else {
-        uct_tcp_ep_destroy_internal(tl_ep);
+        /* notify the peer that nothing will be sent anymore */
+        status = uct_tcp_cm_send_event(ep, UCT_TCP_CM_CONN_FIN, 0);
+        if (status == UCS_OK) {
+            return;
+        }
     }
+
+    uct_tcp_ep_destroy_internal(tl_ep);
```

Comment:
why do we need to call `uct_tcp_ep_destroy_internal` here also when sending FIN?

Reply:
we call `uct_tcp_ep_destroy_internal` in the following cases:
- connect_to_iface
- EP is not connected
- either TX or RX capability isn't set
- sending `FIN` CM packet failed